resteasy-client/src/main/java/org/jboss/resteasy/plugins/providers/sse/client/SseEventSourceScheduler.java

```diff
@@ -0,0 +1,167 @@
+            : scheduledExecutorService;
+      this.shutdownExecutorService = scheduledExecutorService == null;
+      this.phaser = new Phaser(1)
+      {
+         @Override
+         protected boolean onAdvance(int phase, int registeredParties)
+         {
+            boolean terminated = super.onAdvance(phase, registeredParties);
+            if (terminated)
+            {
+               if (onTerminated != null)
+               {
+                  onTerminated.run();
+               }
+            }
+            return terminated;
+         }
+      };
+      this.closed = new AtomicBoolean(false);
+   }
+   
+   void schedule(final Runnable runnable, long delay, TimeUnit unit) throws RejectedExecutionException
+   {
+      if (this.closed.get())
+      {
+         return;
+      }
+      try
+      {
+         this.scheduledExecutorService.schedule(new Runnable()
+         {
+            @Override
+            public void run()
+            {
+               if (SseEventSourceScheduler.this.closed.get())
+               {
+                  return;
+               }
+               int registrationPhase = SseEventSourceScheduler.this.phaser.register();
+               try
+               {
+                  // If this phaser is terminated (registrationPhase < 0)
+                  // (no more registered parties) or the current
+                  // registration phase is other than 0 it means that
+                  // shutdownNow has been called already.
+                  if (registrationPhase != 0)
+                  {
+                     return;
+                  }
+                  runnable.run();
+               }
+               finally
+               {
+                  // We can invoke 'arriveAndDeregister()' safely since
+                  // this method has no effect if the phaser is already
+                  // terminated.
+                  SseEventSourceScheduler.this.phaser.arriveAndDeregister();
+               }
+            }
+         }, delay, unit);
+      }
+      catch (RejectedExecutionException e)
+      {
+         if (this.shutdownExecutorService && this.closed.get())
+         {
+            // At this stage the RejectedExecutionException can be either a
+            // normal consequence of the
+            // 'scheduledExecutorService.shutdownNow(...)' method
+            // invocation and in this case it's not an error at all, or a
+            // real error.
+            // So instead of throwing exception that may not be an
+            // unexpected error at all, it is acceptable to do nothing since
+            // user already asked for shutdown.
+            return;
+         }
+         throw e;
+      }
+   }
+
+   boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException
+   {
+      if (!this.closed.get())
+      {
+         return false;
+      }
+      try
+      {
+         this.phaser.awaitAdvanceInterruptibly(0, timeout, unit);
```

Comment:
If shutdownNow() and phaser.arriveAndDeregister() are executed, this.phaser.awaitAdvanceInterruptibly(0, timeout, unit) always returns immediately. I don't get what's the point of adding this check.

Reply:
Hi @jimma,

The method SseEventSourceScheduler.awaitTermination(long timeout, TimeUnit unit) follows the exact same rule as ExecutorService.awaitTermination(long timeout, TimeUnit unit).
It means that it will return true only if a SseEventSourceScheduler.shutdownNow() has been invoked and if all tasks have completed execution before the timeout occurs.

So in our case suppose that T2 is a running thread executing line https://github.com/resteasy/Resteasy/pull/1677/files/f8a2bdfc85831ad269e514f1de2e9c9bbdf9678c#diff-ef11a0ff038fff5fc91418bec89c4962R97. Since SseEventSourceScheduler.shutdownNow() has not been invoked yet registrationPhase is equal to 0.
So T2 will arrive at line https://github.com/resteasy/Resteasy/pull/1677/files/f8a2bdfc85831ad269e514f1de2e9c9bbdf9678c#diff-ef11a0ff038fff5fc91418bec89c4962R108 but without executing it and will be pause for thread scheduling purpose.
Now T1 invoke SseEventSourceScheduler.shutdownNow() successfully. So phaser now only has 1 party, I mean the one registered by T2. And even if scheduledExecutorService.shutdownNow() has been invoked, it still has a running task: the one running in T2.

So at this step a call to scheduledExecutorService.awaitTermination(long timeout, TimeUnit unit) will never return true since T2 is not resuming.
So SseEventSourceScheduler.awaitTermination(long timeout, TimeUnit unit) should behave the same.
Thanks to phaser that what is done: this.phaser.awaitAdvanceInterruptibly(0, timeout, unit) will never return true since we are still at phase 0 because of the remaining party, the phaser is not terminated yet. It will return true once T2 will resume and finish as required.

Reply:
```
So T2 will arrive at line https://github.com/resteasy/Resteasy/pull/1677/files
/f8a2bdfc85831ad269e514f1de2e9c9bbdf9678c#diff-ef11a0ff038fff5fc91418bec89c4962R108 but without 
executing it and will be pause for thread scheduling purpose.
```
Sorry. I don't understand this.  is  without executing blocked by phraser's unarrived parties or something else ? If I understand correctly, this will be executed without pause.

Reply:
Sorry I was not clear enough. It has nothing to deal with the phaser.
What I mean is:

So T2 will arrive at line https://github.com/resteasy/Resteasy/pull/1677/files/f8a2bdfc85831ad269e514f1de2e9c9bbdf9678c#diff-ef11a0ff038fff5fc91418bec89c4962R108 and lets suppose that the underlying system where the program is running decides to pause this thread at this line to let other threads execute.

Reply:
Hi @rsearls,

Yes this class supports the scenario you are talking about.

In case where scheduledExecutorService is provided, invoking `sseEventSourceScheduler.shutdownNow()` will not shutdown the scheduledExecutorService. So user's tasks in scheduledExecutorService will not be affected at all.

Actually in this case only the sse tasks will be affected:
- The current sse task running (https://github.com/resteasy/Resteasy/pull/1677/files/f8a2bdfc85831ad269e514f1de2e9c9bbdf9678c#diff-ef11a0ff038fff5fc91418bec89c4962R108) will be stopped since `SseEventSourceImpl.close()` close the response stream.
- The sse tasks that are scheduled but not already running will exit as soon as they start running (since the scheduledExecutorService has not been sutted down)  thanks to lines https://github.com/resteasy/Resteasy/pull/1677/files/f8a2bdfc85831ad269e514f1de2e9c9bbdf9678c#diff-ef11a0ff038fff5fc91418bec89c4962R93
and https://github.com/resteasy/Resteasy/pull/1677/files/f8a2bdfc85831ad269e514f1de2e9c9bbdf9678c#diff-ef11a0ff038fff5fc91418bec89c4962R104

Hope it's clearer.

Reply:
@NicoNes, @jimma Great.   Jim I'm onboard for having this merged.